<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Applying Starlng on single-cell data • Starlng</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Applying Starlng on single-cell data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Starlng</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/Starlng.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Core-Bioinformatics/Starlng/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<script src="Starlng_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Applying Starlng on single-cell data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Core-Bioinformatics/Starlng/blob/HEAD/vignettes/Starlng.Rmd" class="external-link"><code>vignettes/Starlng.Rmd</code></a></small>
      <div class="d-none name"><code>Starlng.Rmd</code></div>
    </div>

    
    
<p>In this vignette we will demonstrate how to apply the main Starlng function to single-cell RNA-seq data. We will exemplify this on the PBMC3k dataset.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Core-Bioinformatics/Starlng" class="external-link">Starlng</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SeuratObject</span></span>
<span><span class="co">#&gt; Loading required package: sp</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'SeuratObject'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, t</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.satijalab.org/seurat" class="external-link">SeuratData</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Installed datasets</span> ──────────────────────────────── SeuratData v0.2.2.9002 ──</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">pbmc3k</span> 3.1.4</span></span>
<span><span class="co">#&gt; ────────────────────────────────────── Key ─────────────────────────────────────</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Dataset loaded successfully</span></span>
<span><span class="co">#&gt; <span style="color: #BBBB00;">❯</span> Dataset built with a newer version of Seurat than installed</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">❓</span> Unknown version of Seurat installed</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/packageDescription.html" class="external-link">packageVersion</a></span><span class="op">(</span><span class="st">"Starlng"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] '1.0.0'</span></span></code></pre></div>
<p>We repeated the procedures of downloading, qc filtering, normalising and dimensionality reduction as described in the <code>ClustAssess</code> <a href="https://github.com/Core-Bioinformatics/ClustAssess/blob/main/vignettes/stability-pipeline-description.Rmd" class="external-link">vignette</a>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/InstallData.html" class="external-link">InstallData</a></span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The following packages are already installed and will not be</span></span>
<span><span class="co">#&gt; reinstalled: pbmc3k</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/UpdateSeuratObject.html" class="external-link">UpdateSeuratObject</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span></span>
<span><span class="co">#&gt; Validating object structure</span></span>
<span><span class="co">#&gt; Updating object slots</span></span>
<span><span class="co">#&gt; Ensuring keys are in the proper structure</span></span>
<span><span class="co">#&gt; Warning: Assay RNA changing from Assay to Assay</span></span>
<span><span class="co">#&gt; Ensuring keys are in the proper structure</span></span>
<span><span class="co">#&gt; Ensuring feature names don't have underscores or pipes</span></span>
<span><span class="co">#&gt; Updating slots in RNA</span></span>
<span><span class="co">#&gt; Validating object structure for Assay 'RNA'</span></span>
<span><span class="co">#&gt; Object representation is consistent with the most current Seurat version</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">pbmc3k</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span>, col.name <span class="op">=</span> <span class="st">"percent.mito"</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">pbmc3k</span>, pattern <span class="op">=</span> <span class="st">"^RP[SL][[:digit:]]"</span>, col.name <span class="op">=</span> <span class="st">"percent.rp"</span><span class="op">)</span></span>
<span><span class="va">all.index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">MT.index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"^MT-"</span>, x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">RP.index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"^RP[SL][[:digit:]]"</span>, x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="va">pbmc3k</span><span class="op">[</span><span class="op">!</span><span class="op">(</span><span class="op">(</span><span class="va">all.index</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">MT.index</span><span class="op">)</span> <span class="op">|</span> <span class="op">(</span><span class="va">all.index</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">RP.index</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">pbmc3k</span>, <span class="va">nFeature_RNA</span> <span class="op">&lt;</span> <span class="fl">2000</span> <span class="op">&amp;</span> <span class="va">nCount_RNA</span> <span class="op">&lt;</span> <span class="fl">2500</span> <span class="op">&amp;</span> <span class="va">percent.mito</span> <span class="op">&lt;</span> <span class="fl">7</span> <span class="op">&amp;</span> <span class="va">percent.rp</span> <span class="op">&gt;</span> <span class="fl">7</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">pbmc3k</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The `slot` argument of `SetAssayData()` is deprecated as of SeuratObject 5.0.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use the `layer` argument instead.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">Seurat</span> package.</span></span>
<span><span class="co">#&gt;   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/satijalab/seurat/issues&gt;</span>.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span>
<span><span class="co">#&gt; Warning: The `slot` argument of `GetAssayData()` is deprecated as of SeuratObject 5.0.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use the `layer` argument instead.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">Seurat</span> package.</span></span>
<span><span class="co">#&gt;   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/satijalab/seurat/issues&gt;</span>.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">pbmc3k</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, nfeatures <span class="op">=</span> <span class="fl">3000</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">pbmc3k</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">pbmc3k</span>,</span>
<span>    npcs <span class="op">=</span> <span class="fl">30</span>,</span>
<span>    approx <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">pbmc3k</span>,</span>
<span>    reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>    dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>,</span>
<span>    n.neighbors <span class="op">=</span> <span class="fl">30</span>,</span>
<span>    min.dist <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>    metric <span class="op">=</span> <span class="st">"cosine"</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span></span>
<span><span class="co">#&gt; To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span></span>
<span><span class="co">#&gt; This message will be shown once per session</span></span>
<span><span class="va">pbmc3k</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 13607 features across 2401 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (13607 features, 3000 variable features)</span></span>
<span><span class="co">#&gt;  3 layers present: counts, data, scale.data</span></span>
<span><span class="co">#&gt;  2 dimensional reductions calculated: pca, umap</span></span></code></pre></div>
<p>Applying <code>Starlng</code> is straightforward, using the <code><a href="../reference/starlng_write_app_default.html">starlng_write_app_default()</a></code> function. The function requires a normalised expression matrix, a metadata dataframe, a PCA and UMAP embedding. The function allows controlling the number of points used in inferring the pseudotime trajectory (<code>learn_graph_parameters</code>), filtering the genes based on autocorrelation (<code>gene_filtering_function</code>) and the parameters used in the stability assessment of gene clustering (<code>clustering_parameters</code>).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/starlng_write_app_default.html">starlng_write_app_default</a></span><span class="op">(</span></span>
<span>    folder_path <span class="op">=</span> <span class="st">"starlng_app_pbmc3k"</span>,</span>
<span>    expression_matrix <span class="op">=</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">pbmc3k</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, slot <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span>,</span>
<span>    metadata_df <span class="op">=</span> <span class="va">pbmc3k</span><span class="op">@</span><span class="va">meta.data</span>,</span>
<span>    pca_embedding <span class="op">=</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Embeddings.html" class="external-link">Embeddings</a></span><span class="op">(</span><span class="va">pbmc3k</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span>,</span>
<span>    umap_embedding <span class="op">=</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Embeddings.html" class="external-link">Embeddings</a></span><span class="op">(</span><span class="va">pbmc3k</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span>,</span>
<span>    app_title_name <span class="op">=</span> <span class="st">"Starlng PBMC3K Example"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The function will create a Starlng Shiny application in the specified folder. You can run the application using the following command:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/shiny/man/runApp.html" class="external-link">runApp</a></span><span class="op">(</span><span class="st">"starlng_app_pbmc3k"</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Europe/Bucharest</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] pbmc3k.SeuratData_3.1.4 SeuratData_0.2.2.9002   Seurat_5.3.0           </span></span>
<span><span class="co">#&gt; [4] SeuratObject_5.2.0      sp_2.2-0                Starlng_1.0.0          </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] deldir_2.0-4           pbapply_1.7-4          gridExtra_2.3         </span></span>
<span><span class="co">#&gt;   [4] rlang_1.1.6            magrittr_2.0.4         RcppAnnoy_0.0.22      </span></span>
<span><span class="co">#&gt;   [7] spatstat.geom_3.6-0    matrixStats_1.5.0      ggridges_0.5.7        </span></span>
<span><span class="co">#&gt;  [10] compiler_4.5.1         png_0.1-8              systemfonts_1.3.1     </span></span>
<span><span class="co">#&gt;  [13] vctrs_0.6.5            reshape2_1.4.4         stringr_1.5.2         </span></span>
<span><span class="co">#&gt;  [16] crayon_1.5.3           pkgconfig_2.0.3        fastmap_1.2.0         </span></span>
<span><span class="co">#&gt;  [19] promises_1.3.3         rmarkdown_2.30         ragg_1.5.0            </span></span>
<span><span class="co">#&gt;  [22] purrr_1.1.0            xfun_0.53              cachem_1.1.0          </span></span>
<span><span class="co">#&gt;  [25] jsonlite_2.0.0         goftest_1.2-3          later_1.4.4           </span></span>
<span><span class="co">#&gt;  [28] spatstat.utils_3.2-0   irlba_2.3.5.1          parallel_4.5.1        </span></span>
<span><span class="co">#&gt;  [31] cluster_2.1.8.1        R6_2.6.1               ica_1.0-3             </span></span>
<span><span class="co">#&gt;  [34] spatstat.data_3.1-9    stringi_1.8.7          bslib_0.9.0           </span></span>
<span><span class="co">#&gt;  [37] RColorBrewer_1.1-3     reticulate_1.43.0      spatstat.univar_3.1-4 </span></span>
<span><span class="co">#&gt;  [40] parallelly_1.45.1      lmtest_0.9-40          jquerylib_0.1.4       </span></span>
<span><span class="co">#&gt;  [43] scattermore_1.2        Rcpp_1.1.0             iterators_1.0.14      </span></span>
<span><span class="co">#&gt;  [46] knitr_1.50             tensor_1.5.1           future.apply_1.20.0   </span></span>
<span><span class="co">#&gt;  [49] zoo_1.8-14             sctransform_0.4.2      httpuv_1.6.16         </span></span>
<span><span class="co">#&gt;  [52] Matrix_1.7-4           splines_4.5.1          igraph_2.2.0          </span></span>
<span><span class="co">#&gt;  [55] tidyselect_1.2.1       abind_1.4-8            yaml_2.3.10           </span></span>
<span><span class="co">#&gt;  [58] spatstat.random_3.4-2  spatstat.explore_3.5-3 codetools_0.2-19      </span></span>
<span><span class="co">#&gt;  [61] miniUI_0.1.2           listenv_0.9.1          lattice_0.22-5        </span></span>
<span><span class="co">#&gt;  [64] tibble_3.3.0           plyr_1.8.9             shiny_1.11.1          </span></span>
<span><span class="co">#&gt;  [67] S7_0.2.0               ROCR_1.0-11            evaluate_1.0.5        </span></span>
<span><span class="co">#&gt;  [70] Rtsne_0.17             future_1.67.0          fastDummies_1.7.5     </span></span>
<span><span class="co">#&gt;  [73] desc_1.4.3             survival_3.8-3         polyclip_1.10-7       </span></span>
<span><span class="co">#&gt;  [76] fitdistrplus_1.2-4     pillar_1.11.1          KernSmooth_2.23-26    </span></span>
<span><span class="co">#&gt;  [79] foreach_1.5.2          plotly_4.11.0          generics_0.1.4        </span></span>
<span><span class="co">#&gt;  [82] RcppHNSW_0.6.0         ggplot2_4.0.0          scales_1.4.0          </span></span>
<span><span class="co">#&gt;  [85] globals_0.18.0         xtable_1.8-4           glue_1.8.0            </span></span>
<span><span class="co">#&gt;  [88] lazyeval_0.2.2         tools_4.5.1            data.table_1.17.8     </span></span>
<span><span class="co">#&gt;  [91] RSpectra_0.16-2        RANN_2.6.2             fs_1.6.6              </span></span>
<span><span class="co">#&gt;  [94] dotCall64_1.2          cowplot_1.2.0          grid_4.5.1            </span></span>
<span><span class="co">#&gt;  [97] tidyr_1.3.1            nlme_3.1-168           patchwork_1.3.2       </span></span>
<span><span class="co">#&gt; [100] cli_3.6.5              rappdirs_0.3.3         spatstat.sparse_3.1-0 </span></span>
<span><span class="co">#&gt; [103] textshaping_1.0.4      spam_2.11-1            viridisLite_0.4.2     </span></span>
<span><span class="co">#&gt; [106] dplyr_1.1.4            uwot_0.2.3             gtable_0.3.6          </span></span>
<span><span class="co">#&gt; [109] sass_0.4.10            digest_0.6.37          progressr_0.17.0      </span></span>
<span><span class="co">#&gt; [112] ggrepel_0.9.6          htmlwidgets_1.6.4      farver_2.1.2          </span></span>
<span><span class="co">#&gt; [115] htmltools_0.5.8.1      pkgdown_2.1.3          lifecycle_1.0.4       </span></span>
<span><span class="co">#&gt; [118] httr_1.4.7             mime_0.13              MASS_7.3-65</span></span></code></pre></div>
</div>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Andi Munteanu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
